### DISCUSSION TITLE
Deployment and build pipeline

### DISCUSSION THREAD

**Discussion by GarrisonJ** [digitallyinduced/ihp#1121] [General]
I want to deploy to AWS or some similar service. 

Does anyone have a good system they want to share? 

I don't understand a lot about nix (or docker even), but I want to build a pipeline that builds and deploys for me.

**Comment by mpscholten** [comment]
@Montmorency maybe you can share your approach :)

**Comment by Montmorency** [comment]
Hey sure I can explain a little of how I'm going about it at the moment. 
I have a local build machine running nixos and I use a ec2 instance with a nixos ami on aws. 

Locally I have a build and deploy script in an IHP project that looks like this:

```
nix-build
nix-copy-closure -s admin@....amazonaws.com ./result
make static/prod.css
make static/prod.js
scp -r ./Config admin@...amazonaws.com:/home/admin/app
scp ./static/* admin@...amazonaws.com:/home/admin/app
scp ./Application/*sql admin@...amazonaws.com:/home/admin/app
```

Where the app is set up to run in `/home/admin/app` on the aws server.

@zacwood9 [has a good write up](https://zacwood.me/posts/deploying-ihp/) on the way to configure the nix build script and your nixos config file. 

Once the nix app closure is copied to the prod server you can rebuild the nixos image on the server `nixos-rebuild switch`  and the app running as a systemd process. 

There is another approach where you build the whole system closure locally and push that to the server as well but I haven't tried that yet.
